helix-term/src/ui/picker.rs

```diff
@@ -400,7 +424,7 @@ impl Ord for PickerMatch {
     }
 }
 
-type PickerCallback<T> = Box<dyn Fn(&mut Context, &T, Action)>;
+type PickerCallback<T> = Box<dyn Fn(&mut Context, Option<&T>, Action)>;
```

Comment:
I'm split on the `Option` here for cancel/close vs a separate `cancel_fn` on the `Picker`. The `Option` seems ok and makes sense for the case when you hit Enter with no selection (empty picker) but it makes the diff really noisy - all existing pickers need to change - and the None variant is only handled with the new picker. Plus the None for cancel feels a little awkward: you couldn't distinguish between cancel and hitting Enter with an empty picker in the future if that's something we would want to do.

We could minimize the diff by using the separate `cancel_fn` approach and instead of adding that to the constructor for all `Picker`s, it could be added on to just this picker with a separate `with_close_fn` function (see `Prompt::with_line` for something similar).

What do you think @pascalkuthe? I'm split on these approaches so I defer to whatever you think is more idiomatic

Reply:
The problem with having two separate closures is that you can not share state between them so for example if you create some expensive allocation and then want to do something with it on the callback you need to use some awkward `Arc`.  Now that `let .. else` is stable (great suggestion) I think just adding a short `let .. else` statement at the start of every picker is not that disruptive and would make the diff reasonable to review I think.

An alternative could be to add a trait instead of a second callback:

```
trait PickerAction{
	fn cancel(..);
	fn accept(..);
}

impl<F: Fn(..)>> Picker for F{
    fn accept(&self, ..){
		self(...)
	}
}
```

This allows splitting up the functions while still making it possible to share state between them and by implementing the trait for a closure you don't need to change existing call sites. This option would work but seems a little overkill to me. Using either a `PickerResult` enum (to distinguish accepting an empty picker) or an `Option` together with `let .. else` doesn't feel to disruptive and any existing PRs that add new pickers would only need to add a single `let .. else` line. 

> you couldn't distinguish between cancel and hitting Enter with an empty picker in the future if that's something we would want to do.

Interesting I hadn't considered that. I can't really think of a use-case. Did you have something in mind? We could change this to an appropriate `enum` then. With `let .. else` you would only need to replace `let Some(` with `let PickerResult::Accept(` then (or whatever we call the enum/variant.

Reply:
Good points! I agree about the trait - it does seem a little overkill. I didn't have anything specific in mind with cancel vs. close so we can probably ignore that use-case for now. Let's keep the Option change as it is currrently 👍